Add products data store

louwie17 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Adds a new products data store that makes use of the products API -> https://woocommerce.github.io/woocommerce-rest-api-docs/#list-all-products
For now this only allows for retrieving/search for products and getting the product count.

Partially addresses #32524 

### How to test the changes in this Pull Request:

1. Load this branch, create a couple products and go to **WooCommerce > Home**
2. Open your console and run `wp.data.select('wc/admin/products').getProducts({ status: 'publish'}, undefined);`
3. It should return `undefined` at first
4. Run it again, assuming the request finished, and it should return all the published products. You can try the above again with other items in the query, like `_fields: ['id']` for example, which should only return the `ids`, any query param defined here should work: https://woocommerce.github.io/woocommerce-rest-api-docs/#list-all-products
5. Now do the same for the `getProductsTotalCount` function, by running: `wp.data.select('wc/admin/products').getProductsTotalCount({ status: 'publish'}, undefined);`, this might run the correct number right away if you recently ran `getProducts` with the same query params, otherwise it should return `undefined` intiially.
6. If you have the [Redux Devtools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en) extension, you can look for the `wc/admin/products` store and make sure the state is updated correctly and the actions are triggered correctly.
7. Run `wp.data.select('wc/admin/items').getItems('categories', { });` in your console to see if the `items` store still works correctly, and `wp.data.select('wc/admin/items').getItemsTotalCount('categories', { });`.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm nx changelog <project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

louwie17 committed:
Add products data store

louwie17 committed:
Update selectors

botwoo commented:
:bar_chart: Test reports for d68a8da747b9bc60466769fd5031def86ae797bf `Update selectors` have been published and are accessible through the following links:

- [REST API Test Report](https://woocommerce.github.io/woocommerce-test-reports/pr/32964/api/)
- E2E Test Report (TBC)

Visit the [WooCommerce Test Reports](https://woocommerce.github.io/woocommerce-test-reports) homepage to view all published reports. See the [FAQs](https://github.com/woocommerce/woocommerce-test-reports/wiki/FAQs) page if you're having problems accessing them.

joshuatf commented on packages/js/data/src/products/types.ts:
```diff
@@ -0,0 +1,66 @@
+	status: 'any' | 'draft' | 'pending' | 'private' | 'publish' | 'future';
+	type: 'simple' | 'grouped' | 'external' | 'variable';
+	sku: string;
+	featured: string;
```
Is this supposed to be a `boolean`?

louwie17 replied:
Yup it should be, thanks

joshuatf commented on packages/js/data/src/onboarding/selectors.ts:
```diff
@@ -12,7 +12,7 @@ import {
 	ExtensionList,
 	ProfileItemsState,
-	Product,
+	OnboardingProductType,
```
Good call on this rename!

louwie17 replied:
Yeah, `Product` seemed like a bad name.

joshuatf commented on packages/js/data/src/products/action-types.ts:
```diff
@@ -0,0 +1,8 @@
+export enum TYPES {
+	SET_PRODUCT = 'SET_PRODUCT',
```
For all new data stores, I think we should be using the `REQUEST|SUCCESS|ERROR` patterns.  For example - https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/data/src/countries/action-types.ts

louwie17 replied:
Yeah that makes sense, I debated on that, but didn't want to write all the extra actions haha
But it does seem like the best pattern.

louwie17 replied:
I end up just sticking with `SUCCESS` and `ERROR`, as we don't really need `REQUEST` given we have the resolutions for that, and can use `isResolutionFinished`.

joshuatf commented on packages/js/data/src/products/resolvers.ts:
```diff
@@ -0,0 +1,67 @@
+
+type ProductGetResponse = Product[] | ( { data: Product[] } & Response );
+
+function* request( query: Partial< ProductQuery > ) {
```
This looks very similar to the `items` data store request method.  Any opportunities to dry up some of this logic or are they too unique?

louwie17 replied:
They should be near identical aside from the types, but I can make use of a generic for that.

joshuatf commented on packages/js/data/src/products/types.ts:
```diff
@@ -0,0 +1,66 @@
+	shipping_class: string;
+	attribute: string;
+	attribute_term: string;
+	tax_class: string;
```
This should be an enum, right?

louwie17 replied:
I am not sure where I can find the `enum` values for this, the item schema has it listed as a `string` -> https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/rest-api/Controllers/Version1/class-wc-rest-products-v1-controller.php#L1939-L1942

joshuatf replied:
The REST API states it should be `standard`, `reduced-rate` or `zero-rate`.

https://woocommerce.github.io/woocommerce-rest-api-docs/#list-all-products

joshuatf commented on packages/js/data/src/products/types.ts:
```diff
@@ -0,0 +1,66 @@
+	on_sale: boolean;
+	min_price: string;
+	max_price: string;
+	stock_status: string;
```
Also an enum I believe.

louwie17 replied:
Yeah, I believe you are right, I think just `instock` or `outofstock`

joshuatf commented on packages/js/data/src/products/utils.ts:
```diff
@@ -0,0 +1,37 @@
+ * @param {Object} query Query for product totals count.
+ * @return {string} Resource name for product totals.
+ */
+export function getTotalProductCountResourceName(
```
Nice method!

joshuatf reviewed (commented):
Great work, @louwie17!

Overall this looks very solid.  Left a few minor comments, but let me know if you have any questions.

louwie17 committed:
Update action names, consolidate a request function, and update types

louwie17 commented:
@joshuatf thanks for the review, I have addressed all your feedback in https://github.com/woocommerce/woocommerce/pull/32964/commits/9e316ab67daf5eb64c478021f16e5a9913336ed6, if you have time for another review?

louwie17 commented:
@joshuatf I did add one extra step to the testing instructions for double checking the `items` store.

joshuatf commented on packages/js/data/src/products/resolvers.ts:
```diff
@@ -0,0 +1,48 @@
+		return items;
+	} catch ( error ) {
+		yield getProductsError( query, error );
+		return error;
```
We might want to `throw error` here so that users can catch on `getProducts()`.

`getProducts(...).catch(...)`.

joshuatf commented on packages/js/data/src/products/resolvers.ts:
```diff
@@ -0,0 +1,48 @@
+		yield getProductsTotalCountSuccess( query, totalCount );
+		return totalCount;
+	} catch ( error ) {
+		yield getProductsTotalCountError( query, error );
```
Same comment about throwing here.

joshuatf commented on packages/js/data/src/products/test/reducer.ts:
```diff
@@ -0,0 +1,138 @@
+				},
+			},
+			productsCount: {
+				'total-guyisms:{}': 2,
```
😆

louwie17 replied:
Lol, this was deff not me haha I copy pasted most of these tests and updated them, I somehow never noticed this.

joshuatf commented on packages/js/data/src/products/types.ts:
```diff
@@ -0,0 +1,58 @@
+	date_created_gmt: string;
+	date_modified: string;
+	date_modified_gmt: string;
+	type: 'simple' | 'grouped' | 'external' | 'variable';
```
These can also be extended by hooks/filters on the backend.

joshuatf reviewed (commented):
Hey @louwie17, looks really great.  Sorry, I missed a couple small items my first pass.  I left a couple comments, but let me know if you have any questions.

louwie17 committed:
Make sure we throw errors within resolvers

louwie17 committed:
Update Product types, and allow for more extensibility

louwie17 commented:
@joshuatf thanks again for you review and suggestions, I addressed the PR feedback if you have time for a re-review?

joshuatf reviewed (dismissed):
Tested various parameters and fetching of totals and all worked as expected.  Also double checked to categories for any regresions.

This is a fantastic addition to our data stores.  Thank you for all your hard work on this, @louwie17!  :shipit:

louwie17 committed:
Remove unused action

louwie17 committed:
Make sure we always use id for receiving products and update reducer to handle _fields

louwie17 commented:
@joshuatf I just updated this branch with the fixes for the things we discussed in Slack.
Made sure we always add `id` when retrieving all the products, and updated the reducer to not overwrite and delete fields that are already present (but only overwrite the new one's).

joshuatf reviewed (approved):
Thanks for fixing those up, @louwie17!  Looks good and I'm able to retrieve individual fields without passing `id` now.
